Helm is inconsistent about failed deployments

This is basically a generalized version of the _bug_ #2257, possibly other filed issues and a failure mode that I've encountered today.

It seems that helm does not care if _for some unknown class of reasons_ a deployment fails. Helm will report the failure but then happily false-successfully re-deploy the same chart, without changing the actual state but registering a new state in a history config map (revision). 

There are two ways to trigger helm's schizophrenia:

1. #2257
1a. Install a chart that contains a property that is yet unknown to helm, e.g. `envFrom` before helm v2.4.0: true success
1b. Upgrade the release with a change in its chart that may or may not be related to that property, thus triggering the patch code path: failure
1c. Upgrade the release again without any change: the cluster is in an undefined state and helm lies about it. :(

2. 2a. Install a chart that is both valid and fully supported by helm. My case includes a ClusterRoleBinding: deployment is successful, obviously.
2b. Change the ClusterRoleBinding using an invalid {kind: ServiceAccount, subjects.name:}. Deployment fails as it should.
2c. Deploy the same chart again w/o any changes since 2b. This deployment falsely succeeds and helm is in the same schizophrenic state as in case 1. 

/cc @tback @jayme-github

IIUC, some variant of 2b.  we see Helm/Tiller get out of sync with K8s actual state. It's not envFrom, and it's not silently immutable fields in Deployment or Daemonset.  It has occurred in several different cases when the K8S API server rejecting a patches due to failed validations.  Helm seems to barrel on assuming that the patch succeeded, and recording the desired state as actual.

For example, from the Tiller log:

```
2017/05/16 01:22:37 client.go:251: error updating the resource "service-proxy":
         Deployment.apps "service-proxy" is invalid: spec.template.spec.containers[0].ports[1].protocol: Unsupported value: "HTTP": supported values: TCP, UDP
...
2017/05/16 01:22:38 release_server.go:329: warning: Upgrade "RELEASENAME" failed: Deployment.apps "service-proxy" is invalid: spec.template.spec.containers[0].ports[1].protocol: Unsupported value: "HTTP": supported values: TCP, UDP
2017/05/16 01:22:38 storage.go:59: Updating "RELEASENAME" (v47) in storage
2017/05/16 01:22:38 storage.go:51: Create release "RELEASENAME" (v48) in storage
```

After that, there were some other changes -- in this case I believe to the "image" and "env" of one of the containers -- that didn't occur, but were apparently recorded as being successful. Future upgrades after correcting the protocol to 'TCP' did not update the image and env contents to the new values.

I don't have a minimal test case put together yet, but I do have the full Tiller log and the "helm get" of the Tiller release configmaps from before and after the failure.  

This occurred with Tiller v2.4.1, Helm client v2.3.1 on K8S 1.6.3.

--- a/pkg/storage/storage.go
+++ b/pkg/storage/storage.go
@@ -18,6 +18,7 @@ package storage // import "k8s.io/helm/pkg/storage"
 
 import (
 	"fmt"
+	"strings"
 
 	rspb "k8s.io/helm/pkg/proto/hapi/release"
 	relutil "k8s.io/helm/pkg/releaseutil"
@@ -127,14 +128,13 @@ func (s *Storage) Deployed(name string) (*rspb.Release, error) {
 		"OWNER":  "TILLER",
 		"STATUS": "DEPLOYED",
 	})
-	switch {
-	case err != nil:
-		return nil, err
-	case len(ls) == 0:
-		return nil, fmt.Errorf("%q has no deployed releases", name)
-	default:
+	if err == nil {
 		return ls[0], nil
 	}
+	if strings.Contains(err.Error(), "not found") {
+		return nil, fmt.Errorf("%q has no deployed releases", name)
+	}
+	return nil, err
 }
 
 // History returns the revision history for the release with the provided name, or

--- a/pkg/tiller/release_update.go
+++ b/pkg/tiller/release_update.go
@@ -69,8 +69,8 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
 		return nil, nil, errMissingChart
 	}
 
-	// finds the non-deleted release with the given name
-	currentRelease, err := s.env.Releases.Last(req.Name)
+	// finds the deployed release with the given name
+	currentRelease, err := s.env.Releases.Deployed(req.Name)
 	if err != nil {
 		return nil, nil, err
 	}
@@ -80,9 +80,15 @@ func (s *ReleaseServer) prepareUpdate(req *services.UpdateReleaseRequest) (*rele
 		return nil, nil, err
 	}
 
+	// finds the non-deleted release with the given name
+	lastRelease, err := s.env.Releases.Last(req.Name)
+	if err != nil {
+		return nil, nil, err
+	}
+
 	// Increment revision count. This is passed to templates, and also stored on
 	// the release object.
-	revision := currentRelease.Version + 1
+	revision := lastRelease.Version + 1
 
 	ts := timeconv.Now()
 	options := chartutil.ReleaseOptions{
@@ -151,7 +157,6 @@ func (s *ReleaseServer) performUpdate(originalRelease, updatedRelease *release.R
 	if err := s.ReleaseModule.Update(originalRelease, updatedRelease, req, s.env); err != nil {
 		msg := fmt.Sprintf("Upgrade %q failed: %s", updatedRelease.Name, err)
 		s.Log("warning: %s", msg)
-		originalRelease.Info.Status.Code = release.Status_SUPERSEDED
 		updatedRelease.Info.Status.Code = release.Status_FAILED
 		updatedRelease.Info.Description = msg
 		s.recordRelease(originalRelease, true)

--- a/pkg/tiller/release_update_test.go
+++ b/pkg/tiller/release_update_test.go
@@ -234,8 +234,8 @@ func TestUpdateReleaseFailure(t *testing.T) {
 	if err != nil {
 		t.Errorf("Expected to be able to get previous release")
 	}
-	if oldStatus := oldRelease.Info.Status.Code; oldStatus != release.Status_SUPERSEDED {
-		t.Errorf("Expected SUPERSEDED status on previous Release version. Got %v", oldStatus)
+	if oldStatus := oldRelease.Info.Status.Code; oldStatus != release.Status_DEPLOYED {
+		t.Errorf("Expected Deployed status on previous Release version. Got %v", oldStatus)
 	}
 }
 

